modules/core/src/lifecycle/component-state.ts

```diff
@@ -19,74 +19,70 @@
 // THE SOFTWARE.
 
 import {isAsyncIterable} from '../utils/iterable-utils';
-import {PROP_SYMBOLS} from './constants';
-const {ASYNC_ORIGINAL, ASYNC_RESOLVED, ASYNC_DEFAULTS} = PROP_SYMBOLS;
+import {ASYNC_ORIGINAL, ASYNC_RESOLVED, ASYNC_DEFAULTS} from './constants';
+import type Component from './component';
+import {ComponentProps, StatefulComponentProps} from './component';
+import {PropType} from './prop-types';
 
 const EMPTY_PROPS = Object.freeze({});
 
-export default class ComponentState {
-  constructor(component = null) {
+type AsyncPropState = {
+  type: PropType;
+  lastValue: any;
+  resolvedValue: any;
+  pendingLoadCount: number;
+  resolvedLoadCount: number;
+};
+
+export default class ComponentState<T extends ComponentProps> {
```

Comment:
For me it would really help readability below if we didn't call this `T` but something like `Props`, `PropsT` or `PropsType`.

Seeing things like `keyof T` and `StatefulComponentProps<T>` a couple of pages down makes me scrach my head :)